src/core/dappletContext.ts

```diff
@@ -20,20 +21,70 @@ export class DappletContext {
 
     constructor(config: ContextConfig = DEFAULT_CONFIG) {
         config = { ...DEFAULT_CONFIG, ...config };
-        this.dappletProviders = config.providers!;
+        this.dappletProviders = config.providers! || [];
         this.featureRegistry = new FeatureRegistry(config.features || []);
     }
 
     async processRequest(request: DappletRequest): Promise<DappletTxResult> {
+        // dapplet loading and validation
+        for (let i = 0; i < request.frames.length; i++) {
+            const dapplet = await this._loadDapplet(request.frames[i].dappletId);
+            this._validateDapplet(dapplet);
+            request.frames[i].dapplet = dapplet;
+        }
+
         const engine = new DappletEngine(request, this);
-        engine.onStatusChanged((statuses) => console.log(statuses.map(s => FrameStatus[s]).join(", ")));
-        await engine.load();
-        engine.validate();
-        
+        engine.run();
+
         //const activity = new DappletActivity(request, this);
         return {};
     }
 
+    private async _loadDapplet(dappletId: string): Promise<DappletTemplate> {
+        for (const provider of this.dappletProviders) {
+            try {
+                const dapplet = await provider.loadDapplet(dappletId);
+                if (!dapplet) continue;
+                return dapplet;
+            } catch (err) { }
+        }
+        throw Error(`All configured providers don't contain the dapplet ${dappletId}.`);
+    }
+
+    private _validateDapplet(dapplet: DappletTemplate) {
+        const incompatibleFeatures: string[][] = [];
+
+        // validate and init views
+        let isCompatibleViewFound = false;
+        // ToDo: think about whose view is priority (wallet developer vs dapplet developer)
+        for (const viewTemplate of dapplet.views) {
+            const regKeys = dapplet.aliases[viewTemplate.type];
+            if (!regKeys) throw new Error(`Alias ${viewTemplate.type} is not defined in usings.`);
+            const viewClass = this.featureRegistry.get(regKeys);
```

Comment:
reqKeys is hier an Array of different view aliases. Not the compound key of the single view!
It is a bug.
please introduce `RegKey` as a type for registry keys instead of `string[]` to prevent bugs like this in the future.